[FIX] account,l10n_ke_edi_tremol: send proforma when missing legal info

agr-odoo opened the pull request:
With Kenya localization installed
Create an invoice
Send&Print

Issue: The sytem will issue the final pdf before the invoice has been send to the fiscal device, so the legal information is actually missing A solution is to generate a proforma pdf invoice when the legal info is still missing

Removing call to `_hook_invoice_document_after_pdf_report_render` as it seems intended to add a legal layer not needed in proforma invoice

opw-3599869

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/142861).

agr-odoo commented:
@smetl 
Hello, I had a few ideas on how to fix this, but you may have a better one:
- Just block send&print, seems not user friendly
- Send the proforma invoice instead of the legal one. This is the currently implemented solution
- Send the invoice for legal validation via `_call_web_service_before_invoice_pdf_render`, like we do for other localization (l10n_mx_edi). It doesn't seems easy to implement here because for Kenya the client needs to send the validation request to the IoT device
https://www.odoo.com/documentation/16.0/applications/finance/fiscal_localizations/kenya.html#sending-the-data-to-kra-using-the-tremol-g03-control-unit
https://github.com/odoo/odoo/blame/14cd403acdd39ab64572e0dc352b2d3ac30e0d6e/addons/l10n_ke_edi_tremol/models/account_move.py#L229

lordkrandel commented:
Shouldn't it be a warning before pressing "Send" ? 
And what about multi-document? If you're trying to send 3 invoices, from what I understand the user gets blocked on the first, and the message doesn't actually say which invoice is lacking info...
🤔

agr-odoo commented:
> Shouldn't it be a warning before pressing "Send" ?
It's a possible solution, yes, but can we issue a warning without blocking the operation? Or do you meant to block the action?

>  And what about multi-document?
Currently the idea would be to send proforma invoice when needed and multi-document is actually not working as expected because it just log the message in chatter without generating the proforma :/
If we want to go this route, I need to work a bit more on this

lordkrandel commented:
> Shouldn't it be a warning before pressing "Send" ?
> It's a possible solution, yes, but can we issue a warning without blocking the operation? Or do you meant to block the action?

Warnings appear *before* clicking send.

Just for context, I come from this experience: 
I'm currently refactoring the warnings for Italy which got too long and complicated to just make a big list in the multiple invoices mode  - by dividing issues in "categories"  (fix settings, fix company, fix partners, fix moves)
![image](https://github.com/odoo/odoo/assets/1665365/4e8d491a-ef6e-45ce-8b89-4b9dd420ce6a)

But I was thinking something **much simpler** about letting the user know he's only sending a proforma.
I think we need Laurent or a PO's advice on this... ?

agr-odoo commented:
> Warnings appear _before_ clicking send.

Now I get it, thanks, I'm not that expert of the new Send&Print widget
I added the warning and fixed the multi-document send

lordkrandel commented on addons/l10n_ke_edi_tremol/models/account_move_send.py:
```diff
@@ -0,0 +1,43 @@
+                ):
+                    warning_msg += f"{invoice.name} : " + _(
+                        "The document does not have details related to the fiscal device, a proforma invoice will be sent.") + "\n"
+            if wizard.send_mail_warning_message:
```
I have some doubt on `send_mail_warning_message` because that relates to the Email checkbox and email sending process, which is not the Kenyan flow I guess. 

Here is a reference to my WIP for Italy... I'm not certain this is the way to go for Kenya but just to show you some context: https://github.com/odoo/odoo/pull/142596/files#diff-a93d4442f33dead3e87d7e0e4ef601e1aa14bd48d6810555cf3354a5003c3fe1L10

Thing is, every EDI that needs sending expands this wizard... PEPPOL, Italy, Kenya, I guess more - we can't all just use a single warning field.
Adding another warning computed field would require showing it in a view - this would require the user to upgrade the module.

agr-odoo commented:
> The problem is most likely there in all stables no ?
@clbr-odoo Not really, from saas-16.4 it was introduced the final (official) invoice feature: when the user send&print, the invoice will be considered final and it won't be recreated (unless the original one is deleted). We should ensure invoicing data is complete before generating the final pdf.
In older version the system would recreate the invoice to send each time, so this problem does not exists

aliyatastemirova commented on addons/account_edi_ubl_cii/models/account_move_send.py:
```diff
@@ -124,6 +124,10 @@ def _hook_invoice_document_after_pdf_report_render(self, invoice, invoice_data):
         # EXTENDS 'account'
         super()._hook_invoice_document_after_pdf_report_render(invoice, invoice_data)
 
+        # Skip proforma pdf post processing
```
I can see that in further versions, we use `proforma_pdf_attachment_values` when reading the pdf content below
in 17.0:
```python
pdf_values = invoice_data.get('pdf_attachment_values') or invoice_data['proforma_pdf_attachment_values']
reader_buffer = io.BytesIO(pdf_values['raw'])
 ```

so I don't think this would be the right solution here :thinking:

clbr-odoo commented:
@agr-odoo Ok I see thanks.
Then if that's fine by you I'll push a commit with some proposed changes on this PR (that include fix of problem @aliyatastemirova is raising)
If you agree we can squash them and go for final review with Laurent (las) :slightly_smiling_face:

agr-odoo commented:
@clbr-odoo 
Sure, I can't wait to see this closed :D

clbr-odoo commented:
@agr-odoo 
So I cherry-picked one of juvr's commit from 17.0 that fixes the `_hook_invoice_document_after_pdf_report_render` that was only designed for regular pdf (and not proforma one).

I adapted a bit your work starting from there, obviously they will only have the warning displayed if they `-u l10n_ke_edi_tremol` but at least they get the proforma correctly.

Tell me if/when I can squash our two commits :slightly_smiling_face: 
PS: I allowed myself to resolve conversations to be sure I wasn't missing something.

If any one else has something more to say, now is the time @aliyatastemirova @lordkrandel :pray: 

cc: @smetl

agr-odoo commented on addons/account/wizard/account_move_send.py:
```diff
@@ -597,11 +597,10 @@ def _generate_invoice_fallback_documents(self, invoices_data):
         for invoice, invoice_data in invoices_data.items():
-            form = invoice_data['_form']
-            if form._need_invoice_document(invoice) and invoice_data.get('error'):
+            if self._need_invoice_document(invoice) and invoice_data.get('error'):
```
why reverting this change?
It is needed for multi-record send

clbr-odoo replied:
Strictly speaking I don't think it's necessary, if you look at `_need_invoice_document` and `_hook_invoice_document_after_pdf_report_render` they could both be switched to `@api.model` methods (which we have in 17.0 for all hooks).

But I get your point, will do it for consistency and possible (even if very unlikely) future overrides in 16.4. :thinking: :+1:

agr-odoo replied:
In multi mode `self.invoice_mode` is `invoice_multi` and we don't get `proforma_pdf_attachment` data filled
https://github.com/odoo/odoo/blob/b78412af4e3157fa81bd03514cf5f14517e7a7f8/addons/account/wizard/account_move_send.py#L297
while `form.invoice_mode` is `invoice_single`
But now that I look at it a bit more, it seems a bit hack-ish, perhaps it was supposed to be this way :thinking:

clbr-odoo replied:
Oh indeed, it was definitely hacky, we removed all that form overhead in next version :)

agr-odoo reviewed (commented):
@clbr-odoo 
Thanks for your work, I've tested and everything looks good to me (except for the note I've left)

smetl commented on addons/account_edi_ubl_cii/models/account_move_send.py:
```diff
@@ -145,7 +145,8 @@ def _hook_invoice_document_after_pdf_report_render(self, invoice, invoice_data):
 
         # Read pdf content.
-        reader_buffer = io.BytesIO(invoice_data['pdf_attachment_values']['raw'])
+        pdf_values = invoice_data.get('pdf_attachment_values') or invoice_data['proforma_pdf_attachment_values']
```
Why? In case of error, you don't enter this hook so `proforma_pdf_attachment_values` can't be there :thinking:

clbr-odoo replied:
As discussed you do enter the hook from generate_invoice_fallback :P

smetl commented on addons/l10n_ke_edi_tremol/wizard/account_move_send.py:
```diff
@@ -0,0 +1,34 @@
+
+    def _hook_invoice_document_before_pdf_report_render(self, invoice, invoice_data):
+        super()._hook_invoice_document_before_pdf_report_render(invoice, invoice_data)
+        if invoice.country_code == 'KE' and not invoice._l10n_ke_check_invoice_ready():
```
```suggestion
        if not invoice._l10n_ke_check_invoice_ready():
```
the check of the country is already in `_l10n_ke_check_invoice_ready`
Btw, the name of the method doesn't mean anything... what about `_l10n_ke_is_invoice_edi_needed`

clbr-odoo replied:
Yesn't, if the invoice is not from Kenya at all you don't want to set the error/warning.
Maybe it's clearer if I remove the country check from the method ?

smetl commented on addons/l10n_ke_edi_tremol/wizard/account_move_send.py:
```diff
@@ -0,0 +1,34 @@
+        # sent to the fiscal device yet.
+        self.ensure_one()
+        if all(move.country_code == 'KE' for move in self.move_ids):
+            allow_fallback_pdf = True
```
I don't like it.
- first, check the country of the company instead of each move separately.
- second, such behavior must be managed by the hooks. You are free to put an error or not and let the pdf be generated or not but I really want to prevent overrides of `action_send_and_print` itself

smetl commented on addons/l10n_ke_edi_tremol/wizard/account_move_send.py:
```diff
@@ -0,0 +1,34 @@
+        self.ensure_one()
+        if all(move.country_code == 'KE' for move in self.move_ids):
+            allow_fallback_pdf = True
+        return super().action_send_and_print(from_cron, allow_fallback_pdf)
```
removing explicit keywords for optional parameters is terrorism...

smetl commented on addons/l10n_ke_edi_tremol/tests/test_account_move_send.py:
```diff
@@ -0,0 +1,50 @@
+    @classmethod
+    def setUpClass(cls, chart_template_ref=None):
+        super().setUpClass(chart_template_ref=chart_template_ref)
+        cls.company_data['company'].account_fiscal_country_id = cls.env.ref('base.ke')
```
put the kenyan chart template as parameter instead of this

clbr-odoo commented:
@smetl done :+1:

smetl commented:
@robodoo r+

robodoo commented:
@agr-odoo @smetl because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

smetl commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

clbr-odoo commented:
@agr-odoo it will most likely crash in FW 17.0. Shouldn't be too hard to adapt but let me know if you want me to handle it or double-check your changes :+1:

agr-odoo commented on addons/l10n_ke_edi_tremol/wizard/account_move_send.py:
```diff
@@ -0,0 +1,36 @@
+                "This document does not have details related to the fiscal device, a proforma invoice will be used."
+            )
+
+    def action_send_and_print(self, from_cron=False, allow_fallback_pdf=False):
+        # EXTENDS account - prevent Send & Print if KE invoices aren't validated and no fallback is allowed.
+        self.ensure_one()
+        if not allow_fallback_pdf \
```
@clbr-odoo 
I know I'm a bit too late, but how are we supposed to have `allow_fallback_pdf = True` here?

clbr-odoo replied:
For example it will be True if called from the e-commerce, which is a flow we don't want to block.
If used directly from Send & Print on invoices Laurent decided that it's better to just raise and not allow the proforma.

agr-odoo replied:
Alright, thanks for the info
